File: ios/Simple MP3/Data/VolumeNormalizer.swift
//
//  VolumeNormalizer.swift
//  Simple MP3
//
//  ReplayGain-based loudness normalization (parity with Android's VolumeNormalizer).
//

import AVFoundation
import Foundation

/// ReplayGain stores a per-track gain (dB) in file tags so players can play every
/// track at a consistent perceived loudness. AVPlayer volume maxes at 1.0, so we only
/// ever attenuate — the usual case since RG track gains are almost always negative.
enum VolumeNormalizer {
    static let trackGainKey = "replaygain_track_gain"

    /// Parse a ReplayGain value like "-6.48 dB" or "3.21" into a Double, or nil.
    static func parseGainDb(_ raw: String?) -> Double? {
        guard let raw, !raw.isEmpty else { return nil }
        var cleaned = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if cleaned.lowercased().hasSuffix("db") {
            cleaned = String(cleaned.dropLast(2)).trimmingCharacters(in: .whitespaces)
        }
        return Double(cleaned)
    }

    /// Linear AVPlayer volume for a track. Returns 1.0 when normalization is off or the
    /// track has no known gain, so un-tagged tracks play at the user's own level.
    static func linearVolume(gainDb: Double?, preampDb: Int, enabled: Bool) -> Float {
        guard enabled, let gainDb else { return 1.0 }
        let totalDb = gainDb + Double(preampDb)
        let linear = pow(10.0, totalDb / 20.0)
        return Float(min(1.0, max(0.0, linear)))
    }

    /// Best-effort ReplayGain track-gain read from an asset's metadata. Handles ID3 TXXX
    /// (`extraAttributes[.info]` == REPLAYGAIN_TRACK_GAIN) and freeform/iTunes atoms.
    static func trackGain(from asset: AVAsset) async -> Double? {
        guard let items = try? await asset.load(.metadata) else { return nil }
        for item in items {
            if let gain = await replayGainValue(from: item) { return gain }
        }
        return nil
    }

    private static func replayGainValue(from item: AVMetadataItem) async -> Double? {
        var matched = false
        if let key = item.key as? String, key.lowercased().contains(trackGainKey) {
            matched = true
        }
        if !matched, let id = item.identifier?.rawValue.lowercased(), id.contains(trackGainKey) {
            matched = true
        }
        if !matched,
           let extra = try? await item.load(.extraAttributes),
           let info = extra[.info] as? String,
           info.lowercased().contains(trackGainKey) {
            matched = true
        }
        guard matched else { return nil }
        let value = try? await item.load(.stringValue)
        return parseGainDb(value)
    }
}

File: android/app/src/main/java/io/karpilabs/simplemp3/data/normalization/VolumeNormalizer.kt
package io.karpilabs.simplemp3.data.normalization

import androidx.annotation.OptIn
import androidx.media3.common.Metadata
import androidx.media3.common.util.UnstableApi
import androidx.media3.extractor.metadata.id3.TextInformationFrame
import androidx.media3.extractor.metadata.vorbis.VorbisComment
import kotlin.math.pow

/**
 * ReplayGain-based loudness normalization.
 *
 * ReplayGain stores a per-track gain (in dB) in file tags so players can play every
 * track at a consistent perceived loudness. Because Android's [android.media.audiofx]
 * chain can only attenuate (ExoPlayer volume maxes at 1.0), we never boost above unity
 * — the usual case since RG track gains are almost always negative.
 */
object VolumeNormalizer {
    private const val RG_TRACK_GAIN_KEY = "REPLAYGAIN_TRACK_GAIN"

    /** Parse a ReplayGain value like "-6.48 dB" or "3.21" into a Double, or null. */
    fun parseGainDb(raw: String?): Double? {
        if (raw.isNullOrBlank()) return null
        // Strip a trailing "dB" (any case) and whitespace, keep leading sign / decimals.
        val cleaned =
            raw
                .trim()
                .removeSuffix("dB")
                .removeSuffix("DB")
                .removeSuffix("db")
                .trim()
        return cleaned.toDoubleOrNull()
    }

    /**
     * Extract the track-gain value (dB) from decoded stream [Metadata]. Handles ID3v2
     * TXXX frames (MP3/AAC) and Vorbis comments (FLAC/OGG/Opus).
     */
    @OptIn(UnstableApi::class)
    fun trackGainFromMetadata(metadata: Metadata): Double? {
        for (i in 0 until metadata.length()) {
            when (val entry = metadata.get(i)) {
                is TextInformationFrame -> {
                    if (entry.description.equals(RG_TRACK_GAIN_KEY, ignoreCase = true)) {
                        val value = entry.values.firstOrNull() ?: entry.value
                        parseGainDb(value)?.let { return it }
                    }
                }
                is VorbisComment -> {
                    if (entry.key.equals(RG_TRACK_GAIN_KEY, ignoreCase = true)) {
                        parseGainDb(entry.value)?.let { return it }
                    }
                }
                else -> {}
            }
        }
        return null
    }

    /**
     * Linear ExoPlayer volume for a track. Returns 1.0 when normalization is off or the
     * track has no known gain, so playback is never quieter than the user's own level
     * for un-tagged tracks.
     */
    fun linearVolume(
        gainDb: Double?,
        preampDb: Int,
        enabled: Boolean,
    ): Float {
        if (!enabled || gainDb == null) return 1f
        val totalDb = gainDb + preampDb
        val linear = 10.0.pow(totalDb / 20.0)
        return linear.coerceIn(0.0, 1.0).toFloat()
    }
}
